### DISCUSSION TITLE
Confused about JIT-compilation of nested functions

### DISCUSSION THREAD

**Discussion by dfdx** [jax-ml/jax#21297] [Q&A] [answered]
Does JAX use cache when JIT-compiling nested functions? Consider this example:

```python
import jax
import jax.numpy as jnp


def outer(a: jax.Array):
    def inner(x: jax.Array):
        print("compiling")
        return x * 2
    # print(f"id(inner) = {id(inner)}")
    # print(f"hash(inner) = {hash(inner)}")
    jitted_inner = jax.jit(inner)
    return jitted_inner(a)

a = jax.random.normal(jax.random.key(0), (3, 4))
outer(a)
```
When calling `outer()`, I'd expect `inner()` to be compiled just once and then re-used from the cache. But in practice text "compiling" is printed on each call. 

Some sources say that `jax.jit()` uses function ID as a cache key, others mention the hash. In any case, I added both (uncomment `print()` statements in the code above), and indeed both values change on every call to `outer()`. 

If Python re-defines nested functions on each call, how do I make JIT cache them?

--------------------

Full output after 3 calls:

```
In [48]: outer(a);
id(inner) = 134661989421488
hash(inner) = 8416374338843
compiling

In [49]: outer(a);
id(inner) = 134662026406512
hash(inner) = 8416376650407
compiling

In [50]: outer(a);
id(inner) = 134661989432864
hash(inner) = 8416374339554
compiling
```

-------------------------

System info:
* Python 3.10.12
* JAX 0.4.26

**Comment by dfm** [comment]
As you've identified here, since `inner` is being redefined in each call to `outer`, it is treated as a new function by JAX, and you get a cache miss. In this particular example, I would move the definition of inner up to the global scope:

```python
def inner(x: jax.Array):
    print("compiling")
    return x * 2

def outer(a: jax.Array):
    jitted_inner = jax.jit(inner)
    return jitted_inner(a)
```

which would work as you intend. To be completely explicit I would probably also move the `jit` decorator to `inner`:

```python
@jax.jit
def inner(x: jax.Array):
    print("compiling")
    return x * 2

def outer(a: jax.Array):
    return inner(a)
```

but the former seems to do the trick as well.

In this simple example, I don't see any reason why you wouldn't want to refactor like this, but I'm not sure how easily this generalizes to your case. Either way, I hope it helps!

**Reply by dfdx** [reply]
Indeed, this code is just a simplified example of real use case. In reality, I'm looking at the implementation of [greedy search](https://github.com/huggingface/transformers/blob/481a95781404e48b1c80940be17e8279dec82fe8/src/transformers/generation/flax_utils.py#L565-L649) by Huggingface, which uses two inner functions `greedy_search_body_fn` and `greedy_search_cond_fn`, both of which are actually closures over model and parameters. These functions are implicitly compiled in `lax.while_loop`, but the performance [isn't great](https://github.com/huggingface/transformers/issues/29410). After a [few experiments][3] I realized the inner functions are actually re-compiled on every call, and so here we are. 

More generally, many JAX/Flax tutorials make use of nested functions ([example][4]). If inner functions are re-defined on each call, does it mean the only proper way to handle them is to apply JIT to the outer-most ones?

[1]: https://github.com/huggingface/transformers/blob/481a95781404e48b1c80940be17e8279dec82fe8/src/transformers/generation/flax_utils.py#L565-L649
[2]: https://github.com/huggingface/transformers/issues/29410
[3]: https://github.com/dfdx/fabrique/blob/main/fabrique/generation.py
[4]: https://flax.readthedocs.io/en/latest/flip/1009-optimizer-api.html#optax-training-step

**Reply by dfm** [reply, marked as answer]
That's what I expected! Yeah, like you say, the usual advice here would be to move the `jit` as high up the stack as a you can. For example, in the flax examples that you link to, the `jit` is applied to the training step, e.g.:

```python
@jax.jit
def outer(...):
  def inner(...):
    ...
  inner(...)
```

in which case `inner` is only compiled once!

But, there are cases where this won't necessarily work (e.g. long compile times, etc.). In that case, maybe you could try converting the closure into a compiled function (at the global level) which takes the relevant parameters as static arguments, which should also lead to a cache hit.

**Reply by dfdx** [reply]
It would require quite some refactoring, but eventually should work. Thank you for clarification!

**Reply by JeffGreen** [reply]
As a follow-up, I'd like to understand why / how functions are designated "new" functions in JAX.  

If I declare two lambdas with identical source code, JAX will consider these to be two different functions even though they produce the same outputs, will have the same JIT representation, etc.  The implication here is that JAX is hashing / caching based not on the functions arguments / content (ie `inspect.getsource()` or related), but instead based on the declaration / Callable id() itself.  In an effort to understand JAX and its design decisions more deeply (and having run into this issue myself!), I'm curious why JAX caches based on function declaration / ID as opposed to content.

Thanks.

**Reply by jakevdp** [reply]
Yes, you're correct: JAX caches the JIT-compilation artifact based on the hash/equality of the function object being compiled, not based on the content of the function. What that means, for example, is that this code will lead to 10 compilations:
```python
for i in range(10):
  jax.jit(lambda x: x)(x)
```

**Reply by JeffGreen** [reply]
Thanks.  I'm not familiar with the innards of JAX - would it be prohibitively difficult to hash based on content instead  so that we could avoid redundant recompilations as above?

**Reply by jakevdp** [reply]
It might be possible, but you’d have to keep around references to the content of compiled functions that you keep in the hash. That could quickly get expensive, because functions can be long. So you might end up just keeping weak references, but at that point it wouldn’t solve the problem you’re trying to solve, because in an example like the one above the weak reference would go out of scope before the next call, and you wouldn’t benefit from the hashed content.

Also, keep in mind that determining the content of a function requires tracing the function, which is a nonzero cost that the current hashing approach is able to avoid.

**Reply by yashk2810** [reply]
Also note that (as Jake said above), you can define a `__hash__` and `__eq__` for the function/method you pass to jit and we will respect that. In other words, it's much much easier to do this at the user level than the JAX level.

> I'm not familiar with the innards of JAX 

JAX always does the simple unambiguous thing when there is ambiguity involved.

**Reply by JeffGreen** [reply]
This is all very helpful, thanks!

**Reply by MRiabov** [reply]
> Yes, you're correct: JAX caches the JIT-compilation artifact based on the hash/equality of the function object being compiled, not based on the content of the function. What that means, for example, is that this code will lead to 10 compilations:
> 
> ```python
> for i in range(10):
>   jax.jit(lambda x: x)(x)
> ```

Jake, does this mean that for optimal performance one should never use nested functions? I mean, they are super-ubiquitous in jax docs, and it's very surprising. 

Or, if `jit` is very global - I have a codebase for 10k lines with a single `jit` - would it not recompile?

Cheers.

**Reply by pdcook** [reply]
> Also note that (as Jake said above), you can define a `__hash__` and `__eq__` for the function/method you pass to jit and we will respect that. In other words, it's much much easier to do this at the user level than the JAX level.
> 
> > I'm not familiar with the innards of JAX
> 
> JAX always does the simple unambiguous thing when there is ambiguity involved.

I must be misunderstanding something here. I thought this had solved all my problems, but the following snippet shows that the function is still recompiled 10 times despite both `__hash__` and `__eq__` being defined:

```python
import jax

jax.config.update("jax_explain_cache_misses", True)


class StableFunction:
    def __init__(self, fn):
        self.fn = fn

    def __call__(self, *args, **kwargs):
        return self.fn(*args, **kwargs)

    def __hash__(self):
        return 0  # All instances hash the same

    def __eq__(self, other):
        return isinstance(other, StableFunction)


for i in range(10):

    def f(x):
        return x * x

    s = StableFunction(f)

    jax.jit(s)(i)

```

Result:
```
WARNING:2026-03-10 19:35:51,055:jax._src.interpreters.partial_eval:2329: TRACING CACHE MISS at test.py:10:4 (<module>):
  never seen function:
    f id=130840287890944 defined at test.py:5
WARNING:2026-03-10 19:35:51,443:jax._src.interpreters.partial_eval:2329: TRACING CACHE MISS at test.py:10:4 (<module>):
  never seen function:
    f id=130839100001472 defined at test.py:5
  but seen another function defined on the same line; maybe the function is
  being re-defined repeatedly, preventing caching?
WARNING:2026-03-10 19:35:51,464:jax._src.interpreters.partial_eval:2329: TRACING CACHE MISS at test.py:10:4 (<module>):
  never seen function:
    f id=130839099896864 defined at test.py:5
  but seen another function defined on the same line; maybe the function is
  being re-defined repeatedly, preventing caching?
...
< repeated >
...
```

This is using JAX version 0.9.0.1.

EDIT:

It seems to have to do with the scope of the jitted object. If I save each object to a global list somewhere, then everything works as expected. This seems like the wrong way to go about this.

**Reply by jakevdp** [reply]
> It seems to have to do with the scope of the jitted object.

Exactly – the jit cache is a weakref cache, so as to avoid keeping alive callables that would otherwise be cleaned-up by the Python garbage collector. If you want to avoid cleanup of ephemeral functions, you'll need to do that via some other mechanism.

> This seems like the wrong way to go about this.

What would you suggest?
